### ISSUE TITLE
Producer tests failing with non-beta package 

### ISSUE THREAD

**Issue by bertracoon** [pact-foundation/pact-net#530]
**Previous issues**
I have searched issues in this and other pact-foundation repos, as well as searching online and attempting to sign up for slack.

**Version information:**
 - OS: MacOS M1 ARM (also reproduced on Windows x64 and Azure Devops ubuntu agents. 
 - PactNet Version: 5.0.0
 - .Net Version: 8.0.403
 - Pact Broker Version (if applicable): Consumers using same .NET/PactNet version. Pacts in pactflow.io

**Describe the bug**
Producer tests which previously worked using 5.0.0-beta.2 fail to run when updated to 5.0.0. Tests fail with 

> Request Failed - builder error for url (message://localhost:49152/pact-messages)

**Steps To Reproduce**
Update PactNet from 5.0.0-beta.2 to 5.0.0
Running tests using several test runners and xunit. 

Example test
`using PactNet;
using PactNet.Infrastructure.Outputters;
using PactNet.Output.Xunit;
using PactNet.Verifier;
using System.Text.Json;
using System.Text.Json.Serialization;
using Domain.Events;
using Xunit.Abstractions;

namespace ContractTests;

public class CustomerProductAccessRequestsProviderTests
{
    private static readonly JsonSerializerOptions Options =
    new JsonSerializerOptions
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        WriteIndented = true,
        Converters = { new JsonStringEnumConverter() },
    };

    private readonly PactVerifierConfig _config;

    public CustomerProductAccessRequestsProviderTests(ITestOutputHelper output)
    {
        _config = new PactVerifierConfig
        {
            Outputters = new List<IOutput>
        {
            new XunitOutput(output),
            new ConsoleOutput()
        },

            LogLevel = PactLogLevel.Debug,
        };
    }

    [Fact]
    public void ProviderForConsumerMembershipCreditHold()
    {
        using var pactVerifier = new PactVerifier("CustomerProductAccessBlockRequested", _config);

        pactVerifier
            .WithMessages(scenarios =>
            {
                var message = new CustomerProductAccessBlockRequested (12345,"customerAccountCode");
                scenarios.Add("A CustomerProductAccessBlockRequested message", builder => builder.WithContent(() => message, Options));

            }, Options)
            .WithPactBrokerSource(new Uri(PactBrokerConfig.PactBrokerUrl), options =>
            {
                options.TokenAuthentication(PactBrokerConfig.PactBrokerToken);

                options.PublishResults(PactBrokerConfig.ProviderVersion, publishOptions =>
                {
                    publishOptions.ProviderBranch(PactBrokerConfig.ProviderBranch);
                    publishOptions.ProviderTags(PactBrokerConfig.ProviderTags.Split(','));
                });
            })
            .Verify();
    }

}
`
 

**Expected behavior**
Tests should run and pass or fail based on contract using latest non-beta package. 

**Log Output**

> PactNet.Exceptions.PactVerificationFailedException: Pact verification failed
> 
> PactNet.Exceptions.PactVerificationFailedException
> Pact verification failed
>    at PactNet.Verifier.InteropVerifierProvider.Execute()
>    at PactNet.Verifier.PactVerifierSource.Verify()
>    at CustomerProductAccessRequestsProviderTests.ProviderForConsumerMembershipCreditHold() in /Users/user/Source/test/pact/Tests/CustomerProductAccessRequestsProviderTests.cs:line 43
>    at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
>    at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)
> 
> 
> 
> Starting messaging provider at http://localhost:49152/pact-messages/
> Messaging provider successfully started
> Starting verification...
> Pact verification failed
> 
> Verifier Output
> ---------------
> The pact at https://mycompany.pactflow.io/pacts/provider/CustomerProductAccessBlockRequested/consumer/Membership.CreditHold.CustomerProductAccessBlockRequested/pact-version/7eec1c139f9cff34f415ef40e4cfa4c1add732d4 is being verified because the pact content belongs to the consumer version matching the following criterion:
>     * latest version of Membership.CreditHold.CustomerProductAccessBlockRequested from the main branch 'main' (0.0.2)
> 
> Verifying a pact between Membership.CreditHold.CustomerProductAccessBlockRequested and CustomerProductAccessBlockRequested
> 
>   A CustomerProductAccessBlockRequested message (2s 759ms loading, 274ms verification)
>      Given A plan's credit status needs to be set to Suspended
>       Request Failed - builder error for url (message://localhost:49152/pact-messages)
> 
> 
> Failures:
> 
> 1) Verifying a pact between Membership.CreditHold.CustomerProductAccessBlockRequested and CustomerProductAccessBlockRequested Given A plan's credit status needs to be set to Suspended - A CustomerProductAccessBlockRequested message - builder error for url (message://localhost:49152/pact-messages)
> 
> 
> There were 1 pact failures
> 
> Verifier Logs
> -------------
> 2024-12-16T16:27:03.876145Z  INFO ThreadId(01) pact_verifier::pact_broker: Fetching path '' from pact broker
> 2024-12-16T16:27:03.876229Z DEBUG ThreadId(01) pact_verifier::pact_broker: Final broker URL: https://mycompany.pactflow.io/
> 2024-12-16T16:27:03.876361Z DEBUG ThreadId(01) reqwest::connect: starting new connection: https://mycompany.pactflow.io/    
> 2024-12-16T16:27:03.924305Z DEBUG ThreadId(01) hyper_util::client::legacy::connect::http: connecting to 18.172.88.111:443
> 2024-12-16T16:27:03.947033Z DEBUG ThreadId(01) hyper_util::client::legacy::connect::http: connected to 18.172.88.111:443
> 2024-12-16T16:27:04.709953Z DEBUG ThreadId(01) pact_verifier::pact_broker: Link URL is templated
> 2024-12-16T16:27:04.709963Z DEBUG ThreadId(01) pact_verifier::pact_broker: templated URL = https://mycompany.pactflow.io/pacts/provider/{provider}/for-verification
> 2024-12-16T16:27:04.710970Z DEBUG ThreadId(01) pact_verifier::pact_broker: final URL = https://mycompany.pactflow.io/pacts/provider/CustomerProductAccessBlockRequested/for-verification
> 2024-12-16T16:27:04.710994Z  INFO ThreadId(01) pact_verifier::pact_broker: Fetching path '/pacts/provider/CustomerProductAccessBlockRequested/for-verification' from pact broker
> 2024-12-16T16:27:04.710999Z DEBUG ThreadId(01) pact_verifier::pact_broker: Final broker URL: https://mycompany.pactflow.io/pacts/provider/CustomerProductAccessBlockRequested/for-verification
> 2024-12-16T16:27:04.711016Z DEBUG ThreadId(01) hyper_util::client::legacy::pool: reuse idle connection for ("https", mycompany.pactflow.io)
> 2024-12-16T16:27:05.384786Z DEBUG ThreadId(01) pact_verifier::pact_broker: templated URL = https://mycompany.pactflow.io/pacts/provider/CustomerProductAccessBlockRequested/for-verification
> 2024-12-16T16:27:05.385872Z DEBUG ThreadId(01) pact_verifier::pact_broker: final URL = https://mycompany.pactflow.io/pacts/provider/CustomerProductAccessBlockRequested/for-verification
> 2024-12-16T16:27:05.385902Z DEBUG ThreadId(01) pact_verifier::pact_broker: Sending JSON to https://mycompany.pactflow.io/pacts/provider/CustomerProductAccessBlockRequested/for-verification using POST: {"includePendingStatus":false,"consumerVersionSelectors":[]}
> 2024-12-16T16:27:05.385945Z DEBUG ThreadId(01) hyper_util::client::legacy::pool: reuse idle connection for ("https", mycompany.pactflow.io)
> 2024-12-16T16:27:05.751041Z  INFO ThreadId(01) pact_verifier::pact_broker: Fetching path '/pacts/provider/CustomerProductAccessBlockRequested/consumer/Membership.CreditHold.CustomerProductAccessBlockRequested/pact-version/7eec1c139f9cff34f415ef40e4cfa4c1add732d4/metadata/c1tdW2N2XT0xMTY3' from pact broker
> 2024-12-16T16:27:05.751052Z DEBUG ThreadId(01) pact_verifier::pact_broker: Final broker URL: https://mycompany.pactflow.io/pacts/provider/CustomerProductAccessBlockRequested/consumer/Membership.CreditHold.CustomerProductAccessBlockRequested/pact-version/7eec1c139f9cff34f415ef40e4cfa4c1add732d4/metadata/c1tdW2N2XT0xMTY3
> 2024-12-16T16:27:05.751068Z DEBUG ThreadId(01) hyper_util::client::legacy::pool: reuse idle connection for ("https", mycompany.pactflow.io)
> 2024-12-16T16:27:06.651546Z DEBUG ThreadId(01) verify_interaction{interaction="A CustomerProductAccessBlockRequested message"}: pact_verifier: Executing provider states
> 2024-12-16T16:27:06.651601Z  INFO ThreadId(01) verify_interaction{interaction="A CustomerProductAccessBlockRequested message"}: pact_verifier: Running setup provider state change handler 'A plan's credit status needs to be set to Suspended' for 'A CustomerProductAccessBlockRequested message'
> 2024-12-16T16:27:06.651607Z  WARN ThreadId(01) verify_interaction{interaction="A CustomerProductAccessBlockRequested message"}: pact_verifier::callback_executors: State Change ignored as there is no state change URL provided for interaction ec13b619980eaa70fbc255c9f74a456a6e47efcd
> 2024-12-16T16:27:06.651610Z DEBUG ThreadId(01) verify_interaction{interaction="A CustomerProductAccessBlockRequested message"}: pact_verifier: State Change: "ProviderState { name: "A plan's credit status needs to be set to Suspended", params: {} }" -> Ok({})
> 2024-12-16T16:27:06.651623Z  INFO ThreadId(01) verify_interaction{interaction="A CustomerProductAccessBlockRequested message"}: pact_verifier: Running provider verification for 'A CustomerProductAccessBlockRequested message'
> 2024-12-16T16:27:06.651633Z DEBUG ThreadId(01) verify_interaction{interaction="A CustomerProductAccessBlockRequested message"}: pact_verifier: Verifying an asynchronous message (single shot)
> 2024-12-16T16:27:06.651667Z  INFO ThreadId(01) verify_interaction{interaction="A CustomerProductAccessBlockRequested message"}: pact_verifier::provider_client: Sending request to provider at message://localhost:49152/pact-messages/
> 2024-12-16T16:27:06.651670Z DEBUG ThreadId(01) verify_interaction{interaction="A CustomerProductAccessBlockRequested message"}: pact_verifier::provider_client: Provider details = ProviderInfo { name: "CustomerProductAccessBlockRequested", protocol: "message", host: "localhost", port: Some(49152), path: "/pact-messages/", transports: [ProviderTransport { transport: "message", port: Some(49152), path: Some("/pact-messages/"), scheme: Some("message") }] }
> 2024-12-16T16:27:06.651687Z  INFO ThreadId(01) verify_interaction{interaction="A CustomerProductAccessBlockRequested message"}: pact_verifier::provider_client: Sending request HTTP Request ( method: POST, path: /, query: None, headers: Some({"Content-Type": ["application/json"]}), body: Present(145 bytes, application/json) )
> 2024-12-16T16:27:06.651693Z DEBUG ThreadId(01) verify_interaction{interaction="A CustomerProductAccessBlockRequested message"}: pact_verifier::provider_client: body:
> {"description":"A CustomerProductAccessBlockRequested message","providerStates":[{"name":"A plan's credit status needs to be set to Suspended"}]}
> 2024-12-16T16:27:06.651786Z  INFO ThreadId(01) pact_verifier: Publishing verification results back to the Pact Broker
> 2024-12-16T16:27:06.651788Z DEBUG ThreadId(01) pact_verifier: Publishing a failure result to PactBroker(https://mycompany.pactflow.io/, provider_name='CustomerProductAccessBlockRequested')
> 2024-12-16T16:27:06.855401Z  INFO ThreadId(01) pact_verifier::pact_broker: Fetching path '/pacticipants/CustomerProductAccessBlockRequested' from pact broker
> 2024-12-16T16:27:06.855425Z DEBUG ThreadId(01) pact_verifier::pact_broker: Final broker URL: https://mycompany.pactflow.io/pacticipants/CustomerProductAccessBlockRequested
> 2024-12-16T16:27:06.855453Z DEBUG ThreadId(01) reqwest::connect: starting new connection: https://mycompany.pactflow.io/    
> 2024-12-16T16:27:06.857548Z DEBUG ThreadId(01) hyper_util::client::legacy::connect::http: connecting to 18.172.88.111:443
> 2024-12-16T16:27:06.882562Z DEBUG ThreadId(01) hyper_util::client::legacy::connect::http: connected to 18.172.88.111:443
> 2024-12-16T16:27:07.409114Z DEBUG ThreadId(01) pact_verifier::pact_broker: templated URL = https://mycompany.pactflow.io/pacticipants/CustomerProductAccessBlockRequested/branches/{branch}/versions/{version}
> 2024-12-16T16:27:07.410207Z DEBUG ThreadId(01) pact_verifier::pact_broker: final URL = https://mycompany.pactflow.io/pacticipants/CustomerProductAccessBlockRequested/branches/dev/versions/1.0.0
> 2024-12-16T16:27:07.410223Z DEBUG ThreadId(01) pact_verifier::pact_broker: Sending JSON to https://mycompany.pactflow.io/pacticipants/CustomerProductAccessBlockRequested/branches/dev/versions/1.0.0 using PUT: {}
> 2024-12-16T16:27:07.410258Z DEBUG ThreadId(01) hyper_util::client::legacy::pool: reuse idle connection for ("https", mycompany.pactflow.io)
> 2024-12-16T16:27:07.795506Z DEBUG ThreadId(01) pact_verifier::pact_broker: Pushed branch dev for provider version 1.0.0
> 2024-12-16T16:27:07.795747Z  INFO ThreadId(01) pact_verifier::pact_broker: Fetching path '/pacticipants/CustomerProductAccessBlockRequested' from pact broker
> 2024-12-16T16:27:07.795760Z DEBUG ThreadId(01) pact_verifier::pact_broker: Final broker URL: https://mycompany.pactflow.io/pacticipants/CustomerProductAccessBlockRequested
> 2024-12-16T16:27:07.795838Z DEBUG ThreadId(01) reqwest::connect: starting new connection: https://mycompany.pactflow.io/    
> 2024-12-16T16:27:07.796122Z DEBUG ThreadId(01) hyper_util::client::legacy::pool: reuse idle connection for ("https", mycompany.pactflow.io)
> 2024-12-16T16:27:08.450504Z DEBUG ThreadId(01) pact_verifier::pact_broker: templated URL = https://mycompany.pactflow.io/pacticipants/CustomerProductAccessBlockRequested/versions/{version}/tags/{tag}
> 2024-12-16T16:27:08.451396Z DEBUG ThreadId(01) pact_verifier::pact_broker: final URL = https://mycompany.pactflow.io/pacticipants/CustomerProductAccessBlockRequested/versions/1.0.0/tags/latest
> 2024-12-16T16:27:08.451409Z DEBUG ThreadId(01) pact_verifier::pact_broker: Sending JSON to https://mycompany.pactflow.io/pacticipants/CustomerProductAccessBlockRequested/versions/1.0.0/tags/latest using PUT: {}
> 2024-12-16T16:27:08.451440Z DEBUG ThreadId(01) hyper_util::client::legacy::pool: reuse idle connection for ("https", mycompany.pactflow.io)
> 2024-12-16T16:27:08.821906Z DEBUG ThreadId(01) pact_verifier::pact_broker: Pushed tag latest for provider version 1.0.0
> 2024-12-16T16:27:08.821953Z DEBUG ThreadId(01) pact_verifier::pact_broker: Sending JSON to https://mycompany.pactflow.io/pacts/provider/CustomerProductAccessBlockRequested/consumer/Membership.CreditHold.CustomerProductAccessBlockRequested/pact-version/7eec1c139f9cff34f415ef40e4cfa4c1add732d4/metadata/c1tdW2N2XT0xMTY3/verification-results using POST: {"providerApplicationVersion":"1.0.0","success":false,"testResults":[{"exceptions":[{"message":"builder error for url (message://localhost:49152/pact-messages)"}],"interactionDescription":"A CustomerProductAccessBlockRequested message","interactionId":"ec13b619980eaa70fbc255c9f74a456a6e47efcd","success":false}],"verifiedBy":{"clientLanguage":{"name":"pact-net","testFramework":"pact_ffi","version":"5.0.0.0"},"implementation":"Pact-Rust","version":"1.2.5"}}
> 2024-12-16T16:27:08.821990Z DEBUG ThreadId(01) hyper_util::client::legacy::pool: reuse idle connection for ("https", mycompany.pactflow.io)
> 2024-12-16T16:27:09.547844Z  INFO ThreadId(01) pact_verifier: Results published to Pact Broker
> 2024-12-16T16:27:09.547938Z  WARN ThreadId(01) pact_matching::metrics: 
> --snip

**Additional context**
Comparing this to the run which works successfully using beta.2 it appears to be attempting to connect to message://localhost:49152/pact-messages rather than http://](http://localhost:49152/pact-messages/. This appears to be where it fails. 

`pact_verifier::provider_client: Sending request to provider at message://localhost:49152/pact-messages/`
vs
`pact_verifier::provider_client: Sending request to provider at http://localhost:49152/pact-messages/`

Looking at the changes between beta.2 and 5.0.0 the main thing is the underlying ffi package, it does look like something may have changed with the ProviderInfo class in there. 

https://github.com/pact-foundation/pact-net/compare/5.0.0-beta.2...5.0.0

FFI_VERSION="0.4.16" > FFI_VERSION="0.4.23"

If this is the case I'm surprised we are the only people having this issue unless most are doing API verification.

**Comment by zerdush** [comment]
We are having the same issue. The below sample works for beta2 but not in 5.0.0
[Test Consumer-Test Provider.json](https://github.com/user-attachments/files/18169791/Test.Consumer-Test.Provider.json)
`using System.Text.Json;
using PactNet;
using PactNet.Verifier;
using Xunit.Abstractions;

namespace Tests.Pact;

public class SimpleEventProviderPactTest : IDisposable
{
    private readonly PactVerifier _verifier;
    
    public SimpleEventProviderPactTest(ITestOutputHelper testOutputHelper)
    {
        _verifier = new PactVerifier("Test Provider", new PactVerifierConfig()
        {
            LogLevel = PactLogLevel.Debug,
            Outputters =
            [
                new XunitOutput(testOutputHelper)
            ]
        });
    }
    
    [Fact]
    public void EnsureEventApiHonoursPactWithConsumer()
    {
        var defaultSettings = new JsonSerializerOptions
        {
            PropertyNamingPolicy = JsonNamingPolicy.CamelCase
        };
        
        _verifier
            .WithMessages(scenarios =>
            {
                // register the responses to each interaction
                // the descriptions must match those in the pact file(s)
                scenarios
                    .Add("Simple event published", builder =>
                    {
                        builder.WithContent(() => 
                            
                                new
                                {
                                    Id = "event-001"
                                }
                            );
                    });
            }, defaultSettings)
            .WithFileSource(new FileInfo("./pacts/Test Consumer-Test Provider.json"))
          
            .Verify();
    }
    
    public void Dispose()
    {
        // make sure you dispose the verifier to stop the internal messaging server
        GC.SuppressFinalize(this);
        _verifier.Dispose();
    }
}`

**Comment by adamrodger** [comment]
@mefellows I've not looked into this yet but just tagging for visibility because it looks like it could be an FFI issue

**Comment by YOU54F** [comment]
i saw the same,

https://github.com/pact-foundation/pact-workshop-message/blob/ef0a00509fab5c5e9775ac515d38ee32582f86eb/provider-dotnet-kafka/tests/ProviderApiTests.cs#L66

you need to set the http url for the message proxy server that .net spins up.

you shouldn’t need to in the consumer facing dsl but this is a workaround

**Comment by adamrodger** [comment]
I can recreate this locally, and it I suspect the problem is here:

https://github.com/pact-foundation/pact-net/blob/8661b21112b6b9e56ba31315d027b5b69602b08c/src/PactNet/Verifier/PactVerifier.cs#L116

I remember the entire concept of 'transports' appearing in the FFI, and this code branches differently depending on whether you've already added a HTTP transport or not. If you add the `WithHttpEndpoint` call in then that sets up an HTTP transport, then the `WithMessages` call adds a new messaging transport.

However, if `WithMessages` is the only call then that's set to the primary transport, and I suspect that's what causes the problem, and that's why the workaround works.

I'm looking into it, but yeah I suspect the FFI interaction is a bit brittle and/or it changed after I'd initally written that. That would explain why it worked in 5.0.0-beta.2 but broke in 5.0.0, since the FFI was upgraded between those releases:

| PactNet | FFI |
| ------- | --- |
| 5.0.0-beta.2 | 0.4.17 |
| 5.0.0-beta.3 | 0.4.23 |
| 5.0.0        | 0.4.23 |
| 5.0.1        | 0.4.27 |

This issue is still present in FFI 0.4.27, and thus will still be broken in PactNet 5.0.1 as well. It looks like it broke somewhere between FFI 0.4.17 and 0.4.23.

**Comment by adamrodger** [comment]
I've confirmed that this code is unchanged since 5.0.0-beta.1, so it's definitely something in the FFI that changed between `beta.2` and `beta.3`.

I've changed that locally and I can now specify `WithMessages` without having to specify `WithHttpTransport` first - by making the messaging transport always add a HTTP transport first as well, with an extra messaging transport after it. I'm just testing it out though because I don't really understand why that should work (I suspect a frailty in the FFI that I just happen to not be hitting now, but I don't know if it'll cause other issues).

The comments on `pactffi_verifier_set_provider_info` don't really explain how to use it or what it does, and the comments on `pactffi_verifier_add_provider_transport` allude to magic strings (leaving the scheme as empty/null for messaging pacts) so I'm not too surprised that something changed there.

**Comment by adamrodger** [comment]
Yeah, as suspected that change introduces problems - the order that you call `WithMessages` and `WithHttpTransport` is important, and if you call them the other way around then the HTTP interactions will all fail to verify. The only real difference when trying that out is changing the `messages` string to `http` in the code linked above (and shown in more detail below).

The problem is that the API you're verifying runs on one port and the messaging provider runs on another port. The first transport you add (i.e. which of those you call) is important, because that calls `pactffi_verifier_set_provider_info`. The args passed to that become kind of like the defaults by the looks of things (just looking at the FFI code, things coalescse to those values), but that means if the messaging URL is added first then the regular HTTP interactions start getting sent to the messaging provider instead of the real API now that it's marked as `http` instead of `messages`. Obviously that means they all fail.

That means there are 4 different possibilities:

1. **Only `WithHttpProvider`** - not a problem, we just add the one
2. **Only `WithMessages`** - this is currently a problem because it tries to call `messages://localhost` instead of `http://localhost`, which is where the messaging provider is running. If I change that so it has the correct URL scheme (see below) then that only works in this option but breaks other options.
3. **`WithHttpProvider` first, then `WithMessages`** - not a problem because the defaults get set to the real provider, and only the messaging setup requests get sent to the messaging provider, because it's added second via `pactffi_verifier_add_provider_transport`
4. **`WithMessages` first, then `WithHttpProvider`** - this is a big problem. The `pactffi_verifier_set_provider_info` call happens with the URL of the messaging provider, and then the `pactffi_verifier_add_provider_transport` call for the real provider API. This causes the requests to be sent to the messaging provider instead of the real provider, and so they fail.

So as the code currently stands you must always call `WithHttpProvider` in order to set the defaults in the FFI to be correct, which means options (1) and (3) are the only supported ones. Option (2) is what this bug report is trying to do, and that's why it doesn't work. The workaround is effectively to switch to option (3) by adding the extra call, even though it shouldn't be needed.

Option (2) can be made to work locally, which works because even though the defaults are set to the messaging provider, there are no calls to make to the real provider so it works by chance. The code would look like this:

```cs
public IPactVerifier WithMessages(Action<IMessageScenarios> configure, JsonSerializerOptions settings)
{
    Guard.NotNull(settings, nameof(settings));

    if (this.messageTransportAdded)
    {
        throw new InvalidOperationException("Only one messaging endpoint can be added");
    }

    // start an in-proc server which creates the messaging responses
    Uri uri = this.messagingProvider.Start(settings);

    if (!this.providerInitialised)
    {
        // this is the problem branch for option 4

        this.provider.Initialise();
        // this is effectively a dummy HTTP endpoint that doesn't really get used
        //                                               vvvvv
        //                                               vvvvv note how this is HTTP now, not message
        //                                               vvvvv
        this.provider.SetProviderInfo(this.providerName, uri.Scheme, uri.Host, (ushort)uri.Port, uri.AbsolutePath);
        this.providerInitialised = true;
    }

    // we stil have to do this for the messaging interactions to work
    this.provider.AddTransport("message", (ushort)uri.Port, uri.AbsolutePath, null);

    configure(this.messagingProvider.Scenarios);

    this.messageTransportAdded = true;
    return this;
}
```

However, this still breaks option 4. The provider info gets set to the messaging URL, so then when you add a HTTP transport later, requests still get sent to the messaging provider when they shouldn't. The problem is that at this point we don't know the URL for the real provider (or whether there's even going to be one - we may be in option 2 instead of 4) so we have no proper values to use when setting the provider info.

The logs contain this (note how the provider states PUT calls are being made to the messaging provider, which logs an error because it doesn't support that method, and then the verification fails when it sees the 405 response):

```plain
Starting messaging provider at http://localhost:49152/pact-messages/
Messaging provider successfully started
Starting verification...
Body: WARNING: received messaging request with incorrect method: GET
Body: WARNING: received messaging request with incorrect method: GET
Body: WARNING: received messaging request with incorrect method: PUT
Simulating message with description: an event indicating that an order has been created
Metadata: {"contentType":"application/json"}
Body: {"id":1}
Successfully simulated message with description: an event indicating that an order has been created
Pact verification failed

[...]

2025-03-22T16:19:08.625967Z DEBUG ThreadId(01) verify_interaction{interaction="a request to update the status of an order"}: pact_verifier: Verifying interaction V4 Synchronous/HTTP a request to update the status of an order (None)
2025-03-22T16:19:08.629661Z DEBUG ThreadId(01) verify_interaction{interaction="a request to update the status of an order"}: pact_verifier: Executing provider states
2025-03-22T16:19:08.629670Z  INFO ThreadId(01) verify_interaction{interaction="a request to update the status of an order"}: pact_verifier: Running setup provider state change handler 'an order with ID {id} exists' for 'a request to update the status of an order'
2025-03-22T16:19:08.633110Z DEBUG ThreadId(01) verify_interaction{interaction="a request to update the status of an order"}: pact_verifier::provider_client: Sending HTTP Request ( method: POST, path: /, query: None, headers: Some({"Content-Type": ["application/json"]}), body: Present(75 bytes, application/json) ) to state change handler
2025-03-22T16:19:08.633153Z DEBUG ThreadId(01) verify_interaction{interaction="a request to update the status of an order"}: reqwest::connect: starting new connection: http://localhost:5000/    
2025-03-22T16:19:08.633498Z DEBUG ThreadId(01) verify_interaction{interaction="a request to update the status of an order"}: hyper_util::client::legacy::connect::http: connecting to [::1]:5000
2025-03-22T16:19:08.633849Z DEBUG ThreadId(01) verify_interaction{interaction="a request to update the status of an order"}: hyper_util::client::legacy::connect::http: connected to [::1]:5000
2025-03-22T16:19:08.634351Z DEBUG ThreadId(01) verify_interaction{interaction="a request to update the status of an order"}: hyper_util::client::legacy::pool: pooling idle connection for ("http", localhost:5000)
2025-03-22T16:19:08.634372Z DEBUG ThreadId(01) verify_interaction{interaction="a request to update the status of an order"}: pact_verifier::provider_client: State change request: Response { url: "http://localhost:5000/provider-states", status: 200, headers: {"content-length": "0", "date": "Sat, 22 Mar 2025 16:19:08 GMT", "server": "Kestrel"} }
2025-03-22T16:19:08.634398Z DEBUG ThreadId(01) verify_interaction{interaction="a request to update the status of an order"}: pact_verifier: State Change: "ProviderState { name: "an order with ID {id} exists", params: {"id": Number(1)} }" -> Ok({})
2025-03-22T16:19:08.634410Z  INFO ThreadId(01) verify_interaction{interaction="a request to update the status of an order"}: pact_verifier: Running provider verification for 'a request to update the status of an order'
2025-03-22T16:19:08.634421Z DEBUG ThreadId(01) verify_interaction{interaction="a request to update the status of an order"}: pact_verifier: Verifying a HTTP interaction
2025-03-22T16:19:08.634456Z  INFO ThreadId(01) verify_interaction{interaction="a request to update the status of an order"}: pact_verifier::provider_client: Sending request to provider at http://localhost:49152/pact-messages/
2025-03-22T16:19:08.634458Z DEBUG ThreadId(01) verify_interaction{interaction="a request to update the status of an order"}: pact_verifier::provider_client: Provider details = ProviderInfo { name: "Orders API", protocol: "http", host: "localhost", port: Some(49152), path: "/pact-messages/", transports: [ProviderTransport { transport: "http", port: Some(49152), path: Some("/pact-messages/"), scheme: Some("http") }, ProviderTransport { transport: "message", port: Some(49152), path: Some("/pact-messages/"), scheme: None }, ProviderTransport { transport: "http", port: Some(5000), path: Some("/"), scheme: Some("http") }] }
2025-03-22T16:19:08.634464Z  INFO ThreadId(01) verify_interaction{interaction="a request to update the status of an order"}: pact_verifier::provider_client: Sending request HTTP Request ( method: PUT, path: /api/orders/1/status, query: None, headers: Some({"Content-Type": ["application/json"]}), body: Present(12 bytes, application/json) )
2025-03-22T16:19:08.634470Z DEBUG ThreadId(01) verify_interaction{interaction="a request to update the status of an order"}: pact_verifier::provider_client: body:
"Fulfilling"
2025-03-22T16:19:08.634496Z DEBUG ThreadId(01) verify_interaction{interaction="a request to update the status of an order"}: reqwest::connect: starting new connection: http://localhost:49152/    
2025-03-22T16:19:08.634698Z DEBUG ThreadId(01) verify_interaction{interaction="a request to update the status of an order"}: hyper_util::client::legacy::connect::http: connecting to [::1]:49152
2025-03-22T16:19:08.634938Z DEBUG ThreadId(01) verify_interaction{interaction="a request to update the status of an order"}: hyper_util::client::legacy::connect::http: connected to [::1]:49152
2025-03-22T16:19:08.635214Z DEBUG ThreadId(01) verify_interaction{interaction="a request to update the status of an order"}: hyper_util::client::legacy::pool: pooling idle connection for ("http", localhost:49152)
2025-03-22T16:19:08.635223Z DEBUG ThreadId(01) verify_interaction{interaction="a request to update the status of an order"}: pact_verifier::provider_client: Received native response: Response { url: "http://localhost:49152/pact-messages/api/orders/1/status", status: 405, headers: {"content-length": "62", "content-type": "text/plain", "server": "Microsoft-HTTPAPI/2.0", "date": "Sat, 22 Mar 2025 16:19:08 GMT"} }
2025-03-22T16:19:08.635254Z  INFO ThreadId(01) verify_interaction{interaction="a request to update the status of an order"}: pact_verifier::provider_client: Received response: HTTP Response ( status: 405, headers: Some({"content-type": ["text/plain"], "server": ["Microsoft-HTTPAPI/2.0"], "date": ["Sat, 22 Mar 2025 16:19:08 GMT"], "content-length": ["62"]}), body: Present(62 bytes, text/plain) )
2025-03-22T16:19:08.635259Z DEBUG ThreadId(01) verify_interaction{interaction="a request to update the status of an order"}: pact_verifier::provider_client: body:
WARNING: received messaging request with incorrect method: PUT
2025-03-22T16:19:08.635268Z DEBUG ThreadId(01) verify_interaction{interaction="a request to update the status of an order"}: pact_matching: comparing to expected response: HTTP Response ( status: 204, headers: None, body: Missing )
2025-03-22T16:19:08.635276Z DEBUG ThreadId(01) verify_interaction{interaction="a request to update the status of an order"}: pact_matching: expected content type = '*/*', actual content type = 'text/plain'
2025-03-22T16:19:08.635291Z DEBUG ThreadId(01) verify_interaction{interaction="a request to update the status of an order"}: pact_matching: content type header matcher = 'RuleList { rules: [], rule_logic: And, cascaded: false }'
```

The tldr there is:

- We're verifying a sync HTTP interaction (not a messaging one)
- We send a provider state change to `localhost:5000/proviver-states` (the real provider API, so this is correct)
- We send the real request to `localhost:49152/pact-messages/...` (the messaging provider, so this is incorrect, it should be port 5000 and not have the path prefix)
- The messaging provider responds with HTTP 405 because it doesn't support PUT requests)
- The verification fails

**Comment by adamrodger** [comment]
So - tldr 😁 

- The FFI behaviour changed somewhere between 0.4.17 and 0.4.23
- This means the PactNet behaviour changed between `5.0.0-beta.2` and `5.0.0-beta.3`
- Currently, if you want to use `WithMessages` then it must always be preceded by a call to `WithHttpEndpoint`
- If we change it so that you can omit that call when it should be unnecessary (as you're doing above, because there are no HTTP interactions to verify), then in cases where you want to use both of them then the order of calling them matters, and will break in very confusing ways if you get it wrong. So that's not really a sensible fix.

### Next Steps

I'd like so see better docs in the FFI covering the 4 scenarios I've stated above (and there are more when plugins are involved, but currently not). Once I know how it's supposed to be used then I can make the appropriate changes to do that 👍

It may need a new FFI interaction such that you can overwrite the default provider info, so that if a HTTP transport is added after a non-HTTP then we can detect that and set the HTTP one as the new default. I'm not familiar with the FFI code enough to know if that's reasonable though.

There's a potential workaround where we could detect a HTTP transport being added after a messaging one and throw an exception to say that's not allowed currently, but it appears as though that works by complete chance anyway and could fail in other scenarios (such as verifying multiple pacts or something, not sure). I'm not really comfortable doing that right now unless someone could explain how it's definitely fine to use it that way 😁 I'm not convinced that it is, I think it's a fluke that it happens to work.

**Comment by adamrodger** [comment]
I've just remembered that this sounded familiar - it was a problem I was having after the FFI changed to use "transports", and I requested at thet time that this was changed in this issue:

https://github.com/pact-foundation/pact-reference/issues/297

Unfortunately that was over 18 months ago and I don't think anything has changed since then.
